File: 2024/day03/python/solution.py
#!/usr/bin/env python3
import re
from pathlib import Path

def part1(data: str) -> int:
    """Find all valid mul(X,Y) instructions and sum their products."""
    pattern = r'mul\((\d{1,3}),(\d{1,3})\)'
    matches = re.findall(pattern, data)
    return sum(int(x) * int(y) for x, y in matches)

def part2(data: str) -> int:
    """Like part1, but do() enables and don't() disables mul instructions."""
    # Find all mul instructions and do/don't toggles with their positions
    mul_pattern = r'mul\((\d{1,3}),(\d{1,3})\)'
    do_pattern = r'do\(\)'
    dont_pattern = r"don't\(\)"

    total = 0
    enabled = True

    # Build a list of all events with positions
    events = []
    for m in re.finditer(mul_pattern, data):
        events.append((m.start(), 'mul', int(m.group(1)), int(m.group(2))))
    for m in re.finditer(do_pattern, data):
        events.append((m.start(), 'do', 0, 0))
    for m in re.finditer(dont_pattern, data):
        events.append((m.start(), 'dont', 0, 0))

    # Sort by position and process
    events.sort(key=lambda x: x[0])

    for pos, event_type, x, y in events:
        if event_type == 'do':
            enabled = True
        elif event_type == 'dont':
            enabled = False
        elif event_type == 'mul' and enabled:
            total += x * y

    return total

def main():
    input_path = Path(__file__).parent.parent / 'input.txt'
    data = input_path.read_text()

    print('Part 1:', part1(data))
    print('Part 2:', part2(data))

if __name__ == '__main__':
    main()

File: 2024/day03/ruby/solution.rb
#!/usr/bin/env ruby

def part1(data)
  # Find all valid mul(X,Y) instructions and sum their products
  pattern = /mul\((\d{1,3}),(\d{1,3})\)/
  matches = data.scan(pattern)
  matches.sum { |x, y| x.to_i * y.to_i }
end

def part2(data)
  # Like part1, but do() enables and don't() disables mul instructions
  mul_pattern = /mul\((\d{1,3}),(\d{1,3})\)/
  do_pattern = /do\(\)/
  dont_pattern = /don't\(\)/

  total = 0
  enabled = true

  # Build a list of all events with positions
  events = []

  # Find all mul instructions with their positions
  data.to_enum(:scan, mul_pattern).each do
    match = Regexp.last_match
    events << [match.begin(0), 'mul', match[1].to_i, match[2].to_i]
  end

  # Find all do() instructions with their positions
  data.to_enum(:scan, do_pattern).each do
    pos = Regexp.last_match.begin(0)
    events << [pos, 'do', 0, 0]
  end

  # Find all don't() instructions with their positions
  data.to_enum(:scan, dont_pattern).each do
    pos = Regexp.last_match.begin(0)
    events << [pos, 'dont', 0, 0]
  end

  # Sort by position and process
  events.sort_by! { |e| e[0] }

  events.each do |pos, event_type, x, y|
    case event_type
    when 'do'
      enabled = true
    when 'dont'
      enabled = false
    when 'mul'
      total += x * y if enabled
    end
  end

  total
end

def main
  input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
  data = File.read(input_path)

  puts "Part 1: #{part1(data)}"
  puts "Part 2: #{part2(data)}"
end

main if __FILE__ == $PROGRAM_NAME
